File: src/C3.Catalogue/CassetteModels/CassetteModelService.vb
Imports System.Globalization

Namespace CassetteModels

    Public NotInheritable Class CassetteModelService

        Private ReadOnly _repository As ICassetteModelRepository

        Public Sub New(repository As ICassetteModelRepository)
            If repository Is Nothing Then
                Throw New ArgumentNullException("repository")
            End If
            _repository = repository
        End Sub

        Public Function GetAll() As IList(Of CassetteModel)
            Return _repository.GetAll().OrderBy(
                Function(value As CassetteModel) value.Identifier,
                StringComparer.OrdinalIgnoreCase).ToList()
        End Function

        Public Function Find(identifier As String) As CassetteModel
            Return _repository.FindByIdentifier(Normalize(identifier))
        End Function

        Public Function Create(
                draft As CassetteModelDraft,
                addedAt As DateTime) As CassetteModelOperationResult

            Dim validation As CassetteModelOperationResult = ValidateDraft(draft)
            If Not validation.IsSuccess Then
                Return validation
            End If

            Dim brandCode As String = Normalize(draft.BrandCode).ToUpperInvariant()
            If Not _repository.BrandExists(brandCode) Then
                Return CassetteModelOperationResult.Failed(
                    CassetteModelFailure.BrandNotFound,
                    "The selected brand no longer exists.")
            End If

            Dim code As String = Normalize(draft.Code).ToUpperInvariant()
            Dim identifier As String = brandCode &
                draft.TypeNumber.ToString(CultureInfo.InvariantCulture) & code
            If _repository.IdentifierExists(identifier) Then
                Return CassetteModelOperationResult.Failed(
                    CassetteModelFailure.DuplicateIdentifier,
                    "Cassette model identifier '" & identifier & "' is already in use.")
            End If

            Dim value As New CassetteModel(
                brandCode,
                draft.TypeNumber,
                draft.ModelName.Trim(),
                code,
                identifier,
                Normalize(draft.DisplayName),
                0,
                addedAt,
                Normalize(draft.Notes))

            Try
                _repository.Add(value)
                Return CassetteModelOperationResult.Success(value)
            Catch ex As Exception
                Return CassetteModelOperationResult.Failed(
                    CassetteModelFailure.StorageFailure,
                    ex.Message)
            End Try
        End Function

        Public Function Update(
                identifier As String,
                draft As CassetteModelDraft) As CassetteModelOperationResult

            Dim normalizedIdentifier As String = Normalize(identifier)
            Dim existing As CassetteModel = _repository.FindByIdentifier(normalizedIdentifier)
            If existing Is Nothing Then
                Return CassetteModelOperationResult.Failed(
                    CassetteModelFailure.NotFound,
                    "The selected cassette model no longer exists.")
            End If
            If draft Is Nothing OrElse String.IsNullOrWhiteSpace(draft.ModelName) Then
                Return CassetteModelOperationResult.Failed(
                    CassetteModelFailure.ModelNameRequired,
                    "Model name is required.")
            End If

            Dim updated As New CassetteModel(
                existing.BrandCode,
                existing.TypeNumber,
                draft.ModelName.Trim(),
                existing.Code,
                existing.Identifier,
                Normalize(draft.DisplayName),
                existing.TapeCount,
                existing.AddedAt,
                Normalize(draft.Notes))

            Try
                _repository.Update(updated)
                Return CassetteModelOperationResult.Success(updated)
            Catch ex As Exception
                Return CassetteModelOperationResult.Failed(
                    CassetteModelFailure.StorageFailure,
                    ex.Message)
            End Try
        End Function

        Public Function Delete(identifier As String) As CassetteModelOperationResult
            Dim normalizedIdentifier As String = Normalize(identifier)
            Dim existing As CassetteModel = _repository.FindByIdentifier(normalizedIdentifier)
            If existing Is Nothing Then
                Return CassetteModelOperationResult.Failed(
                    CassetteModelFailure.NotFound,
                    "The selected cassette model no longer exists.")
            End If
            If _repository.IsReferencedByTape(normalizedIdentifier) Then
                Return CassetteModelOperationResult.Failed(
                    CassetteModelFailure.ReferencedByTape,
                    "Cassette model '" & normalizedIdentifier & "' cannot be deleted while tapes use it.")
            End If

            Try
                _repository.Delete(normalizedIdentifier)
                Return CassetteModelOperationResult.Success(existing)
            Catch ex As Exception
                Return CassetteModelOperationResult.Failed(
                    CassetteModelFailure.StorageFailure,
                    ex.Message)
            End Try
        End Function

        Private Shared Function ValidateDraft(
                draft As CassetteModelDraft) As CassetteModelOperationResult

            If draft Is Nothing OrElse String.IsNullOrWhiteSpace(draft.BrandCode) Then
                Return CassetteModelOperationResult.Failed(
                    CassetteModelFailure.BrandRequired,
                    "A brand is required.")
            End If
            If draft.TypeNumber < 1 OrElse draft.TypeNumber > 4 Then
                Return CassetteModelOperationResult.Failed(
                    CassetteModelFailure.InvalidType,
                    "Cassette type must be between I and IV.")
            End If
            If String.IsNullOrWhiteSpace(draft.ModelName) Then
                Return CassetteModelOperationResult.Failed(
                    CassetteModelFailure.ModelNameRequired,
                    "Model name is required.")
            End If

            Dim code As String = Normalize(draft.Code).ToUpperInvariant()
            If code.Length <> 2 OrElse Not code.All(
                    Function(character As Char) character >= "A"c AndAlso character <= "Z"c) Then
                Return CassetteModelOperationResult.Failed(
                    CassetteModelFailure.InvalidCode,
                    "Model code must contain exactly two letters (A-Z).")
            End If

            Return CassetteModelOperationResult.Success(Nothing)
        End Function

        Private Shared Function Normalize(value As String) As String
            Return If(value, String.Empty).Trim()
        End Function

    End Class

End Namespace

File: src/C3.Catalogue/CassetteModels/CassetteModelService.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;

namespace C3.Catalogue.CassetteModels
{
    public sealed class CassetteModelService
    {
        private readonly ICassetteModelRepository repository;

        public CassetteModelService(ICassetteModelRepository repository)
        {
            if (repository == null)
            {
                throw new ArgumentNullException(nameof(repository));
            }

            this.repository = repository;
        }

        public IList<CassetteModel> GetAll()
        {
            return repository.GetAll()
                .OrderBy(value => value.Identifier, StringComparer.OrdinalIgnoreCase)
                .ToList();
        }

        public CassetteModel Find(string identifier)
        {
            return repository.FindByIdentifier(Normalize(identifier));
        }

        public CassetteModelOperationResult Create(
            CassetteModelDraft draft,
            DateTime addedAt)
        {
            var validation = ValidateDraft(draft);
            if (!validation.IsSuccess)
            {
                return validation;
            }

            var brandCode = Normalize(draft.BrandCode).ToUpperInvariant();
            if (!repository.BrandExists(brandCode))
            {
                return CassetteModelOperationResult.Failed(
                    CassetteModelFailure.BrandNotFound,
                    "The selected brand no longer exists.");
            }

            var code = Normalize(draft.Code).ToUpperInvariant();
            var identifier = brandCode +
                draft.TypeNumber.ToString(CultureInfo.InvariantCulture) + code;
            if (repository.IdentifierExists(identifier))
            {
                return CassetteModelOperationResult.Failed(
                    CassetteModelFailure.DuplicateIdentifier,
                    "Cassette model identifier '" + identifier + "' is already in use.");
            }

            var value = new CassetteModel(
                brandCode,
                draft.TypeNumber,
                draft.ModelName.Trim(),
                code,
                identifier,
                Normalize(draft.DisplayName),
                0,
                addedAt,
                Normalize(draft.Notes));
            try
            {
                repository.Add(value);
                return CassetteModelOperationResult.Success(value);
            }
            catch (Exception exception)
            {
                return CassetteModelOperationResult.Failed(
                    CassetteModelFailure.StorageFailure,
                    exception.Message);
            }
        }

        public CassetteModelOperationResult Update(
            string identifier,
            CassetteModelDraft draft)
        {
            var normalizedIdentifier = Normalize(identifier);
            var existing = repository.FindByIdentifier(normalizedIdentifier);
            if (existing == null)
            {
                return CassetteModelOperationResult.Failed(
                    CassetteModelFailure.NotFound,
                    "The selected cassette model no longer exists.");
            }

            if (draft == null || string.IsNullOrWhiteSpace(draft.ModelName))
            {
                return CassetteModelOperationResult.Failed(
                    CassetteModelFailure.ModelNameRequired,
                    "Model name is required.");
            }

            var updated = new CassetteModel(
                existing.BrandCode,
                existing.TypeNumber,
                draft.ModelName.Trim(),
                existing.Code,
                existing.Identifier,
                Normalize(draft.DisplayName),
                existing.TapeCount,
                existing.AddedAt,
                Normalize(draft.Notes));
            try
            {
                repository.Update(updated);
                return CassetteModelOperationResult.Success(updated);
            }
            catch (Exception exception)
            {
                return CassetteModelOperationResult.Failed(
                    CassetteModelFailure.StorageFailure,
                    exception.Message);
            }
        }

        public CassetteModelOperationResult Delete(string identifier)
        {
            var normalizedIdentifier = Normalize(identifier);
            var existing = repository.FindByIdentifier(normalizedIdentifier);
            if (existing == null)
            {
                return CassetteModelOperationResult.Failed(
                    CassetteModelFailure.NotFound,
                    "The selected cassette model no longer exists.");
            }

            if (repository.IsReferencedByTape(normalizedIdentifier))
            {
                return CassetteModelOperationResult.Failed(
                    CassetteModelFailure.ReferencedByTape,
                    "Cassette model '" + normalizedIdentifier +
                    "' cannot be deleted while tapes use it.");
            }

            try
            {
                repository.Delete(normalizedIdentifier);
                return CassetteModelOperationResult.Success(existing);
            }
            catch (Exception exception)
            {
                return CassetteModelOperationResult.Failed(
                    CassetteModelFailure.StorageFailure,
                    exception.Message);
            }
        }

        private static CassetteModelOperationResult ValidateDraft(CassetteModelDraft draft)
        {
            if (draft == null || string.IsNullOrWhiteSpace(draft.BrandCode))
            {
                return CassetteModelOperationResult.Failed(
                    CassetteModelFailure.BrandRequired,
                    "A brand is required.");
            }

            if (draft.TypeNumber < 1 || draft.TypeNumber > 4)
            {
                return CassetteModelOperationResult.Failed(
                    CassetteModelFailure.InvalidType,
                    "Cassette type must be between I and IV.");
            }

            if (string.IsNullOrWhiteSpace(draft.ModelName))
            {
                return CassetteModelOperationResult.Failed(
                    CassetteModelFailure.ModelNameRequired,
                    "Model name is required.");
            }

            var code = Normalize(draft.Code).ToUpperInvariant();
            if (code.Length != 2 || code.Any(character => character < 'A' || character > 'Z'))
            {
                return CassetteModelOperationResult.Failed(
                    CassetteModelFailure.InvalidCode,
                    "Model code must contain exactly two letters (A-Z).");
            }

            return CassetteModelOperationResult.Success(null);
        }

        private static string Normalize(string value)
        {
            return (value ?? string.Empty).Trim();
        }
    }
}
